### PR TITLE
Improve syncer stopping

### LINKED ISSUE #455: Crash when syncer is no longer able to connect to a physical cluster
**Describe the bug**
kcp crashes when it is unable to connect to a previously-registered physical cluster.

**To Reproduce**
1. `make build`
2. `./bin/kcp start --install-cluster-controller --install-namespace-scheduler --install-workspace-scheduler --auto-publish-apis --resources-to-sync="ingresses.networking.k8s.io,deployments.apps,services" --push-mode --token-auth-file contrib/demo/workspaceKubectlPlugin-script/kcp-tokens`
3. Start a physical cluster (e.g. kind)
4. Create a cluster.yaml for the physical cluster and apply it
5. Let the cluster be reconciled/synced at least once
6. Shut down the physical cluster
7. Wait for kcp to crash (it doesn't take all that long)

**Expected behavior**
No crash

**Additional context**
Here's the stack trace (testing from commit a7be6c6):

```
I0207 21:51:05.674008   82886 cluster.go:43] reconciling cluster "kind-us-east1"
I0207 21:51:05.680872   82886 syncer.go:165] Set up informer for /v1, Resource=namespaces
I0207 21:51:05.680899   82886 syncer.go:165] Set up informer for /v1, Resource=services
I0207 21:51:05.680911   82886 syncer.go:165] Set up informer for apps/v1, Resource=deployments
I0207 21:51:05.680922   82886 syncer.go:165] Set up informer for networking.k8s.io/v1, Resource=ingresses
E0207 21:51:15.412338   82886 memcache.go:179] couldn't get current server API group list: Get "https://127.0.0.1:53373/api?timeout=32s": read tcp 127.0.0.1:61746->127.0.0.1:53373: read: connection reset by peer
E0207 21:51:15.412364   82886 discovery.go:112] error mapping resource "ingresses.networking.k8s.io": Get "https://127.0.0.1:53373/api?timeout=32s": read tcp 127.0.0.1:61746->127.0.0.1:53373: read: connection reset by peer
E0207 21:51:15.731331   82886 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3192440, 0x50ce440})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:48 +0x75
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
I0207 21:51:15.744231   82886 panic.go:1038] Shutting down Cluster controller
E0207 21:51:15.744304   82886 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3192440, 0x50ce440})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:74 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001333220})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:48 +0x75
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2bde252]

goroutine 3751 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001333220})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc00903dc20})
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/runtime/runtime.go:55 +0xd8
panic({0x3192440, 0x50ce440})
        /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:1038 +0x215
github.com/kcp-dev/kcp/pkg/syncer.(*Controller).Stop(0xc009684080)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:289 +0x32
github.com/kcp-dev/kcp/pkg/syncer.StartSyncer({0x39f04b0, 0xc002f0a080}, 0x6, 0x4, 0xc0090e9590, {0xc00798e7d0, 0xd}, {0xc00798e72b, 0x5}, 0x2)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/syncer/syncer.go:80 +0x1cc
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).reconcile(0xc001726120, {0x39f04b0, 0xc002f0a080}, 0xc0090d46e0)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/cluster.go:142 +0x101e
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).process(0xc001726120, {0x39f04b0, 0xc002f0a080}, {0xc003b336e0, 0x15})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:277 +0x137
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).processNextWorkItem(0xc001726120, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:255 +0x11c
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*Controller).startWorker(...)
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:239
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start.func1()
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x4a
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xbf22438)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:155 +0x67
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002ae75c0, {0x39a1a20, 0xc003ac51d0}, 0x1, 0xc002ae75c0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:156 +0xb6
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc002f0a080, 0x989680, 0x0, 0x80, 0x0)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:133 +0x89
k8s.io/apimachinery/pkg/util/wait.Until(...)
        /Users/angoldst/go/pkg/mod/github.com/kcp-dev/kubernetes/staging/src/k8s.io/apimachinery@v0.0.0-20220203192237-6b61675b7950/pkg/util/wait/wait.go:90
github.com/kcp-dev/kcp/pkg/reconciler/cluster.(*PreparedController).Start(0xc00068fc98, {0x39f04b0, 0xc002f0a080})
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/reconciler/cluster/controller.go:235 +0x21b
created by github.com/kcp-dev/kcp/pkg/server.(*Server).installClusterController.func1
        /Users/angoldst/go/src/github.com/kcp-dev/kcp/pkg/server/controllers.go:252 +0x4e5
 ```

### PR DESCRIPTION
**Pull request by ncdc**
Switch to creating a cancelable context from the outside (i.e. not in
the Syncer) and letting the caller manage the cancel functions.

Make the controller code more idiomatic.

Fixes #455

### CODE REVIEW

File: pkg/reconciler/cluster/controller.go [hunk 1]
```diff
@@ -145,7 +144,7 @@ type Controller struct {
 	kubeconfig                   clientcmdapi.Config
 	resourcesToSync              []string
 	syncerMode                   SyncerMode
-	syncers                      map[string]*syncer.Syncer
+	syncers                      map[string]func()
```
Comment (imjasonh) [inline pkg/reconciler/cluster/controller.go:147]: Maybe comment that this holds cancel funcs? Or rename to `syncerCancelFuncs`?

**Review by imjasonh — APPROVED** [review]

File: pkg/reconciler/cluster/controller.go (hunk 1 above)
Comment (ncdc) [reply pkg/reconciler/cluster/controller.go:147]: Can do

File: pkg/syncer/syncer.go [hunk 2]
```diff
@@ -249,71 +233,52 @@ type holder struct {
...
+	defer c.queue.ShutDown()
 
-	c.fromInformers.Start(c.ctx.Done())
-	c.fromInformers.WaitForCacheSync(c.ctx.Done())
+	c.fromInformers.Start(ctx.Done())
+	c.fromInformers.WaitForCacheSync(ctx.Done())
 
 	for i := 0; i < numThreads; i++ {
-		go c.startWorker(c.ctx)
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
```
Comment (stevekuznetsov) [inline pkg/syncer/syncer.go:248]: What's the motivation for this change?

File: test/e2e/framework/kcp.go [hunk 3]
```diff
@@ -337,7 +335,7 @@ func (c *kcpServer) monitorEndpoint(client *rest.RESTClient, endpoint string) {
 	}
 	wait.UntilWithContext(ctx, func(ctx context.Context) {
 		_, err := rest.NewRequest(client).RequestURI(endpoint).Do(ctx).Raw()
-		if c.ctx.Err() != nil {
+		if errors.Is(err, context.Canceled) {
```
Comment (stevekuznetsov) [inline test/e2e/framework/kcp.go:338]: I'm not sure this is right - if the context controlling the monitor goroutine is cancelled, we should abort doing _anything_. The `errors.Is()` check makes sense on the consumer end, not the producer?

Comment (ncdc) [reply test/e2e/framework/kcp.go:338]: Oh thanks, yeah I missed they were 2 different contexts

Comment (ncdc) [reply test/e2e/framework/kcp.go:338]: How about `if errors.Is(c.ctx.Err(), context.Canceled) {`

Comment (stevekuznetsov) [reply test/e2e/framework/kcp.go]: I don't think we need to care if the context was cancelled or timed out?

Comment (stevekuznetsov) [reply test/e2e/framework/kcp.go]: From the doc:

```
	// If Done is not yet closed, Err returns nil.
	// If Done is closed, Err returns a non-nil error explaining why:
	// Canceled if the context was canceled
	// or DeadlineExceeded if the context's deadline passed.
	// After Err returns a non-nil error, successive calls to Err return the same error.
	Err() error
```

Here we just care to see if it's done, for any reason

Comment (ncdc) [reply test/e2e/framework/kcp.go]: Ok I changed it back but it’s flaking again. I have a theory:
1. ctx is first set to c.ctx
1. It’s then changed to a context w/a deadline
1. c.ctx is NOT updated w/a deadline (appropriate)
1. When the test completes, its cleanup function cancels the deadline context
1. The ctx passed into wait.UntilWithContext is the deadline context
1. Test cleanup happens -> deadline context is canceled
1. rest request returns an error, which is context.Canceled, but c.ctx hasn’t been canceled (yet)

File: pkg/syncer/syncer.go (hunk 2 above)
Comment (ncdc) [reply pkg/syncer/syncer.go:253]: This is consistent with standard controller design. It restarts any crashed/stopped worker goroutines once per second if needed (but ideally they don't crash/stop).

**Review by stevekuznetsov — APPROVED** [review]
